integrations/apify/src/lib/sync-orchestrator.ts

```diff
@@ -0,0 +1,127 @@
+import { DatasetItem, ApifyDataset } from '../misc/types'
+import { DataTransformer } from '../helpers/data-transformer'
+import { BotpressHelper } from '../helpers/botpress-helper'
+import * as bp from '.botpress'
+import { RuntimeError } from '@botpress/sdk'
+
+export class SyncOrchestrator {
+  private dataTransformer: DataTransformer
+  private botpressHelper: BotpressHelper
+
+  constructor(
+    private bpClient: bp.Client,
+    private logger: bp.Logger
+  ) {
+    this.dataTransformer = new DataTransformer(logger)
+    this.botpressHelper = new BotpressHelper(bpClient)
+  }
+
+  async fetchAndSyncStreaming(
+    dataset: ApifyDataset,
+    kbId: string,
+    syncFunction: (items: DatasetItem[], kbId: string) => Promise<number>,
+    _timeLimitMs: number = 0,
+    startOffset: number = 0
+  ): Promise<{ itemsProcessed: number; hasMore: boolean; nextOffset: number; total: number; filesCreated: number }> {
+    const streamStartTime = Date.now()
+    let offset = startOffset
+    let total = 0
+    let itemsProcessed = 0
+    let filesCreated = 0
+
+    // process files one at a time with 100 seconds timeout detection
+    const MAX_EXECUTION_TIME = 100000
```

Comment:
thought: use a setTimeout. we need this to interrupt whatever long-running task his happening, for the safety of the sync